### PR TITLE
fixed conflict check for place direction

### PR DESCRIPTION
**Pull request by WilliamHarmonMS**
@joshlory 

this fixes [issue439](https://github.com/code-dot-org/craft/issues/439)

something broke the actionPlane conflict checking when I ported over everything to placeDirection. So double woops, but I fixed that too.

### CODE REVIEW

File: src/js/game/GameController.js [hunk 1]
```diff
@@ -1182,13 +1183,17 @@ class GameController {
       soundEffect = () => this.levelView.audioPlayer.play("fizz");
     }
 
+
     this.levelView.playPlaceBlockInFrontAnimation(player, player.position, player.facing, position, () => {
-      this.levelModel.placeBlockDirection(blockType, placementPlane, player, direction);
-      this.levelView.refreshGroundGroup();
+      if (!this.checkConflict(position, placementPlane)) {
```
Comment (joshlory) [inline src/js/game/GameController.js]: I think this should happen in the `canPlaceBlockDirection` check in LevelModel?

File: src/js/game/GameController.js [hunk 2]
```diff
@@ -1198,6 +1203,22 @@ class GameController {
     });
   }
 
+  checkConflict(position, placementPlane) {
+    var conflict = false;
+    this.levelEntity.entityMap.forEach( workingEntity => {
+      if (Position.equals(workingEntity.position, position)) {
+        conflict = true;
+      }
+    });
```
Comment (joshlory) [inline src/js/game/GameController.js]: Nit:

```js
const entityConflict = this.levelEntity.entityMap.some(entity => {
  return Position.equals(entity.position, position);
});
```

Comment (WilliamHarmonMS) [reply src/js/game/GameController.js]: Uncaught TypeError: this.levelEntity.entityMap.some is not a function

😭

File: src/js/game/GameController.js (hunk 1 above)
Comment (WilliamHarmonMS) [reply src/js/game/GameController.js]: This is probably true. I will move the logic 👍

File: src/js/game/GameController.js [hunk 3]
```diff
@@ -1182,13 +1182,15 @@ class GameController {
       soundEffect = () => this.levelView.audioPlayer.play("fizz");
     }
 
+
```
Comment (joshlory) [inline src/js/game/GameController.js]: Extra new lines?

File: src/js/game/LevelMVC/LevelModel.js [hunk 4]
```diff
@@ -597,9 +597,29 @@ module.exports = class LevelModel {
         return false;
       }
     }
+
+    if (this.checkConflict(this.getMoveDirectionPosition(entity, direction), plane)) {
+      return false;
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelModel.js]: Would be great to have unit test coverage of this line!

File: src/js/game/LevelMVC/LevelModel.js [hunk 5]
```diff
@@ -597,9 +597,29 @@ module.exports = class LevelModel {
...
+    }
     return this.getPlaneToPlaceOn(this.getMoveDirectionPosition(entity, direction), entity) !== null;
   }
 
+  checkConflict(position, placementPlane) {
+    var conflict = false;
+    this.controller.levelEntity.entityMap.forEach( workingEntity => {
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelModel.js]: Tiny nit: extra space after `forEach(`

File: src/js/game/LevelMVC/LevelModel.js (hunk 4 above)
Comment (WilliamHarmonMS) [reply src/js/game/LevelMVC/LevelModel.js]: Roger that 👍

File: src/js/game/LevelMVC/LevelModel.js [hunk 6]
```diff
@@ -597,9 +597,29 @@ module.exports = class LevelModel {
...
+
+    if (this.checkConflict(this.getMoveDirectionPosition(entity, direction), plane)) {
+      return false;
+    }
     return this.getPlaneToPlaceOn(this.getMoveDirectionPosition(entity, direction), entity) !== null;
   }
 
+  checkConflict(position, placementPlane) {
+    var conflict = false;
+    this.controller.levelEntity.entityMap.forEach ( workingEntity => {
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelModel.js]: Ah sorry, this should be `forEach(workingEntity` — no space before or after the paren.

File: src/js/game/LevelMVC/LevelModel.js [hunk 7]
```diff
@@ -597,9 +597,29 @@ module.exports = class LevelModel {
...
+      if (Position.equals(workingEntity.position, position)) {
+        conflict = true;
+      }
+    });
+
+    if ((this.actionPlane === placementPlane && placementPlane.getBlockAt(position).blockType !== "") ||
+        this.groundDecorationPlane.getBlockAt(position).blockType !== "") {
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelModel.js]: I think this case is already covered by [line 596](https://github.com/code-dot-org/craft/pull/440/files#diff-82ea2716b14e187d2ef963445c025e64R596) above?

**Pull request merged**